### PR TITLE
Some extensions allowing users to pass functions to define units for new particle types

### PR DESCRIPTION
**Pull request by LonelyCat124**
Sort of a workaround more than something massively portable, but it does allow me to add new particle types without needing to edit the files in swiftsimio.

The tests create boundary and fluid particles with some arbitrary units/cosmology.

I'll try and include some examples and document the examples later.

The examples I was using was:
writing_test.py
```
from swiftsimio import Writer
import swiftsimio.metadata.particle as swp
import swiftsimio.metadata.writer.required_fields as swmw
import swiftsimio.metadata.unit.unit_fields as swuf
import numpy as np
import unyt
import pdb


def generate_units(m, l, t, I, T):
    dict_out = swuf.generate_units(m , l, t, I, T)

    boundary = {
            "coordinates" : l,
            "masses": m,
            "particle_ids": None,
            "velocities": l / t,
            "potential": l * l / (t * t),
            "density": m / (l ** 3),
            "entropy": m * l ** 2 / (t ** 2 * T),
            "internal_energy": (l / t) ** 2,
            "smoothing_length": l,
            "pressure": m / (l * t ** 2),
            "diffusion": None,
            "sfr": m / t,
            "temperature": T,
            "viscosity": None,
            "specific_sfr": 1 / t,
            "material_id": None,
            "diffusion": None,
            "viscosity": None,
            "radiated_energy": m * (l / t) ** 2,
            }
    fluid = { **boundary }

    dict_out["boundary"] = boundary
    dict_out["fluid"] = fluid
    return dict_out


#Use default units, i.e. cm, grams, seconds, Ampere, Kelvin
unit_system = unyt.UnitSystem(name="default", length_unit=unyt.cm, mass_unit=unyt.g, time_unit=unyt.s)

#Add boundary particle type
swp.particle_name_underscores[6] = "boundary"
swp.particle_name_class[6] = "Boundary"
swp.particle_name_text[6] = "Boundary"

swp.particle_name_underscores[7] = "fluid"
swp.particle_name_class[7] = "Fluid"
swp.particle_name_text[7] = "Fluid"

swmw.fluid = {'smoothing_length' : 'SmoothingLength', **swmw.shared}
swmw.boundary = { **swmw.fluid }



n_hydro = 1
n_boundary = 6
n_p = n_hydro + n_boundary

boxsize = 100*unyt.cm

x = Writer(unit_system, boxsize, unit_fields_generate_units=generate_units)


x.boundary.coordinates = np.zeros((n_boundary,3)) * unyt.cm
for i in range(0, n_boundary // 2):
    x.boundary.coordinates[i][0] = (45.0 + float(5*i)) * unyt.cm
    x.boundary.coordinates[i][1] = 25.0 * unyt.cm
    x.boundary.coordinates[i][2] = 50.0 * unyt.cm

for i in range(0, n_boundary//2):
    x.boundary.coordinates[n_boundary//2 + i][0] = (45.0 + float(5*i)) * unyt.cm
    x.boundary.coordinates[n_boundary//2 + i][1] = 35.0 * unyt.cm
    x.boundary.coordinates[n_boundary//2 + i][2] = 50.0 * unyt.cm

x.boundary.velocities = np.zeros((n_boundary,3)) * unyt.cm/unyt.s

x.boundary.masses = np.ones(n_boundary, dtype = float) * unyt.g

x.boundary.smoothing_length = np.ones(n_boundary, dtype = float) * (5.0 * unyt.cm)



x.fluid.coordinates = np.zeros((n_hydro,3)) * unyt.cm
x.fluid.coordinates[0][0] = 50.0 * unyt.cm
x.fluid.coordinates[0][1] = 75.0 * unyt.cm
x.fluid.coordinates[0][2] = 50.0 * unyt.cm


x.fluid.velocities = np.zeros((n_hydro,3)) * unyt.cm/unyt.s

x.fluid.masses = np.ones(n_hydro, dtype = float) * unyt.g

x.fluid.smoothing_length = np.ones(n_hydro, dtype = float) * (5.0 * unyt.cm)

x.write("boundary_test.hdf5")
```

reading_test.py
```
import matplotlib.pyplot as plt
import swiftsimio as sw
import swiftsimio.metadata.particle as swp
import swiftsimio.metadata.writer.required_fields as swmw
import swiftsimio.metadata.unit.unit_fields as swuf
import swiftsimio.metadata.cosmology as swcf


def generate_units(m, l, t, I, T):
    dict_out = swuf.generate_units(m , l, t, I, T)

    boundary = {
            "coordinates" : l,
            "masses": m,
            "particle_ids": None,
            "velocities": l / t,
            "potential": l * l / (t * t),
            "density": m / (l ** 3),
            "entropy": m * l ** 2 / (t ** 2 * T),
            "internal_energy": (l / t) ** 2,
            "smoothing_length": l,
            "pressure": m / (l * t ** 2),
            "diffusion": None,
            "sfr": m / t,
            "temperature": T,
            "viscosity": None,
            "specific_sfr": 1 / t,
            "material_id": None,
            "diffusion": None,
            "viscosity": None,
            "radiated_energy": m * (l / t) ** 2,
            }
    fluid = { **boundary }

    dict_out["boundary"] = boundary
    dict_out["fluid"] = fluid
    return dict_out

def generate_cosmology(scale_factor: float, gamma: float):
    from swiftsimio.objects import cosmo_factor, a
    def cosmo_factory(a_dependence):
        return cosmo_factor(a_dependence, scale_factor)

    dict_out = swcf.generate_cosmology(scale_factor, gamma)
    no_cosmology = cosmo_factory(a / a)
    UNKNOWN = no_cosmology

    boundary = { **dict_out["gas"] }
    fluid = { **dict_out["gas"] }

    dict_out["boundary"] = boundary
    dict_out["fluid"] = fluid

    return dict_out


#Add boundary particle type
swp.particle_name_underscores[6] = "boundary"
swp.particle_name_class[6] = "Boundary"
swp.particle_name_text[6] = "Boundary"

swp.particle_name_underscores[7] = "fluid"
swp.particle_name_class[7] = "Fluid"
swp.particle_name_text[7] = "Fluid"

sw.metadata.particle_fields.boundary = { **sw.metadata.particle_fields.gas }
sw.metadata.particle_fields.fluid =  { **sw.metadata.particle_fields.gas }

swmw.fluid = {'smoothing_length' : 'SmoothingLength', **swmw.shared}
swmw.boundary = { **swmw.fluid }


data = sw.load("boundary_test.hdf5", generate_unit_func=generate_units, generate_cosmology_func=generate_cosmology)

print(type(data.boundary.coordinates.value))
#print(data.gas.coordinates.value)
#print(data.gas.coordinates.value[0][:])
#print(data.gas.coordinates.value[1][:])
# [y[0] for y in x]
plt.scatter([y[0] for y in data.boundary.coordinates.value], [y[1] for y in data.boundary.coordinates.value])
plt.scatter([y[0] for y in data.fluid.coordinates.value], [y[1] for y in data.fluid.coordinates.value])


plt.show()
```

### CODE REVIEW

File: swiftsimio/reader.py [hunk 1]
```diff
@@ -624,10 +626,12 @@ class SWIFTDataset(object):
     SWIFTDataset.metadata.
     """
 
-    def __init__(self, filename, mask=None):
+    def __init__(self, filename, mask=None, generate_unit_func=metadata.unit_fields.generate_units,
+                 generate_cosmology_func=metadata.cosmology.cosmology_fields.generate_cosmology):
```
Comment (JBorrow) [inline swiftsimio/reader.py]: Could you add type annotations here for callables? You can do this by writing `generate_unit_func:Callable=metadata.unit_fields.generate_units`

File: swiftsimio/reader.py [hunk 2]
```diff
@@ -624,10 +626,12 @@ class SWIFTDataset(object):
...
         self.filename = filename
         self.mask = mask
-
+        self.generate_unit_func = generate_unit_func
+        self.generate_cosmology_func = generate_cosmology_func
```
Comment (JBorrow) [inline swiftsimio/reader.py]: Add newline here between the if and assignments

File: swiftsimio/writer.py [hunk 3]
```diff
@@ -333,9 +336,9 @@ def _write_metadata(self, handle: h5py.File, names_to_write: List):
         Writes metadata to file based on the information passed to the object
         and the information in the particle groups.
         """
-
-        number_of_particles = [0] * 6
-        mass_table = [0.0] * 6
+        part_types = max(metadata.particle.particle_name_underscores.keys())+1
```
Comment (JBorrow) [inline swiftsimio/writer.py]: Is this okay within SWIFT? I don't like the idea of doing this at all...

Comment (LonelyCat124) [reply swiftsimio/writer.py]: Whats your concern here? I just tested an input i created that had 1 gas particle, 6 particles of type 6  (boundary) and 1 of type 7 (fluid). SWIFT opened the file fine, and imported 1 gas particle (as expected) before giving up in space_regrid due to not enough particles. Anything that is past the end of `swift_type_count` is ignored. For our work we'll probably want to extend/override this but also not remove already reserved part names, so adding on boundary and fluid particles as new types works. This was the only way I could think of extending the number of allowed types without manually just changing the value of 6.

**Comment by LonelyCat124** [comment]
I think most of those should be addressed now.
For a unit test, shall I just create a simple example that writes out and reads in particles that aren't in the metadata. If it can write out and read in the file (and have the correct particles) after then we should be fine?

File: swiftsimio/writer.py (hunk 3 above)
Comment (JBorrow) [reply swiftsimio/writer.py]: Oh I had just assumed that using any number of items here apart from 6 would kill SWIFT -- if this works then I'm happy to go with this.

**Comment by JBorrow** [comment]
Thanks Aidan! This will be on PyPI at some point

**Pull request merged**